Initial port to Swift 4

tcldr opened the pull request:
What's in this pull request?
============================

Initial migration of library to Swift 4. (Minor changes but lib required by Swiftz which has more extensive changes)

Why merge this pull request?
============================

Migrates library to latest version of Swift. All tests passing with the Swift 4 version of SwiftCheck that I submitted as a PR earlier.

What's worth discussing about this pull request?
================================================

This PR focuses on getting the build compiling without warning and passing all tests. I will have missed anything not picked up by these! There may be some wider changes you have in mind to take advantage of Swift 4. 

What downsides are there to merging this pull request?
======================================================

Uses temporary refs to my personal SwiftCheck forks with Swift 4 migration. Will need updating to point to official SwiftCheck repos.

CodaFi committed:
Merge pull request #58 from typelift/swift-develop

Update to Swift 3.0

Commit:
Fix SPM tests layout

CodaFi committed:
Merge pull request #59 from apparition47/spm-test-module-fix

Fix SPM tests layout

CodaFi committed:
Remove products from the package manifest

CodaFi committed:
Merge pull request #60 from CodaFi/call-of-the-postal-horn

Remove products from the package manifest

CodaFi committed:
Update to Swift 3.1

CodaFi committed:
Merge pull request #61 from CodaFi/upper-crust

Update to Swift 3.1

CodaFi committed:
Bump podspec version

CodaFi committed:
Merge pull request #63 from CodaFi/up-var

Bump podspec version

bkase committed:
Make Operadics constraint looser in SwiftPM

Algebra needs at least 0.2.3 now, so this change is needed to use
Algebra and Swiftx in the same project.

bkase committed:
Merge pull request #65 from bkase/satisfy-operadics-easily

Make Operadics constraint looser in SwiftPM

tcldr committed:
Migration to Swift 4.
Updated to use SPM v4 directory layout/file format
Updated for Xcode 9, beta 3
Set Travis to run tests on Linux build

bkase commented:
Alright, I merged the SwiftCheck PR -- so now we can point to that and then we can merge this one!

tcldr commented:
Nice one. I've updated Operadics to SPM v4 too. Once that's in we can cascade the rest up!

CodaFi commented on Package.resolved:
```diff
@@ -0,0 +1,25 @@
+{
```
This file should be ignored

tcldr replied:
Sure. Added to `.gitignore` along with the Cartfile resolved. My understanding was that these files *should* be committed however to ensure that those using the project are using the last known working versions of any dependencies.

CodaFi replied:
Sure, but Operadics is essentially frozen, and the other two are test-only dependencies.  It doesn’t matter which version our own internal repos are built against, it won’t affect our clients.

CodaFi commented on Sources/Operators.swift:
```diff
@@ -0,0 +1,236 @@
+//
```
We should take a dependency on Operadics, not include it directly.

tcldr replied:
Good spot. Not sure how that got in there.

CodaFi commented:
Xcode 9 redid the file paths so they match the project structure.  This is undesirable given that I would like to support just a package build in the future.  Could you please try to keep everything in their original positions (modulo moving things into the target directory)?

CodaFi commented on Cartfile.private:
```diff
@@ -1,3 +1,3 @@
-github "typelift/Operadics" 
-github "typelift/SwiftCheck"
+github "typelift/Operadics" "swift-develop"
```
You may remove these.  I've just pushed Xcode 9 updates to SwiftCheck and Operadics

tcldr committed:
Updated deps to point to Xcode9/Swift 4 master branches
Added *.resolved to .gitignore
Updated travis to use Swift 4 release build

Removed unreferenced Operators file

Cleaned up folder hierarchy

CodaFi commented:
Resolved by the merge of #68
